cmd/lifecycle-util.go

```diff
@@ -0,0 +1,277 @@
+	return bucketName
+}
+
+func getHostCfgFromURL(urlStr string) *hostConfigV9 {
+	clientURL := newClientURL(urlStr)
+	alias := splitStr(clientURL.String(), "/", 3)[0]
+	return mustGetHostConfig(alias)
+}
+
+func getParsedHostURL(urlStr string) (*url.URL, *probe.Error) {
+	hostCfg := getHostCfgFromURL(urlStr)
+	if hostCfg == nil {
+		return nil, errInvalidURL(urlStr)
+	}
+	hostURLParse, err := url.Parse(hostCfg.URL)
+	if err != nil {
+		return nil, probe.NewError(err)
+	}
+	return hostURLParse, nil
+}
+
+func getMinioClient(urlStr string) (*minio.Client, *probe.Error) {
+	var api *minio.Client
+	hostCfg := getHostCfgFromURL(urlStr)
+	if hostCfg == nil {
+		return nil, errInvalidTarget(urlStr)
+	}
+	creds := credentials.NewStaticV4(hostCfg.AccessKey, hostCfg.SecretKey, "")
+	options := minio.Options{
+		Creds:        creds,
+		Secure:       false,
+		Region:       "",
+		BucketLookup: minio.BucketLookupPath,
+	}
+	parsedHostURL, err := getParsedHostURL(urlStr)
+	if err != nil {
+		return nil, err
+	}
+	api, apierr := minio.NewWithOptions(parsedHostURL.Host, &options) // Used as hostname or host:port
+	if apierr != nil || api == nil {
+		return nil, probe.NewError(apierr)
+	}
+	if globalDebug && isatty.IsTerminal(os.Stdout.Fd()) {
+		api.TraceOn(os.Stderr)
+	}
+	return api, nil
+}
+
+func getIlmConfig(urlStr string) (lfcInfo lifecycleConfiguration, err error) {
+	var lfcInfoXML string
+	var pErr *probe.Error
+	if lfcInfoXML, pErr = getIlmInfo(urlStr); pErr != nil {
+		return lfcInfo, pErr.ToGoError()
+	}
+	if lfcInfoXML == "" {
+		console.Infoln("The lifecycle configuration for " + urlStr + " has not been set.")
+		return lfcInfo, nil
+	}
+
+	if err = xml.Unmarshal([]byte(lfcInfoXML), &lfcInfo); err != nil {
+		return lfcInfo, err
+	}
+	return lfcInfo, nil
+}
+
+// Get lifecycle info (XML) from alias & bucket
+func getIlmInfo(urlStr string) (string, *probe.Error) {
+	bucketName := getBucketNameFromURL(urlStr)
+	if bucketName == "" {
+		err := errors.New("bucket not found")
+		fatalIf(probe.NewError(err), "Bucket Not found.")
+	}
+	api, err := getMinioClient(urlStr)
+	if err != nil || api == nil {
+		fatalIf(err, "Unable to call lifecycle configuration operations on "+urlStr)
+	}
+
+	lifecycleInfo, glcerr := api.GetBucketLifecycle(bucketName)
+	var pErr *probe.Error
+	if glcerr != nil {
+		pErr = probe.NewError(glcerr)
+		if _, glcerr = api.GetBucketLocation(bucketName); glcerr != nil {
+			fatalIf(probe.NewError(glcerr), "Unable to access bucket location or bucket lifecycle configuration.")
+		} else {
+			pErr = nil
+		}
+		return "", pErr
+	}
+
+	return lifecycleInfo, nil
+}
+
+func printIlmJSON(info lifecycleConfiguration) {
+	var ilmJSON []byte
+	var err error
+	// mc ilm generate output is also intended to be used to redirect to a file and used with mc ilm set
+	if isatty.IsTerminal(os.Stdout.Fd()) {
+		ilmJSON, err = json.MarshalIndent(info, "", "    ")
+	} else {
+		ilmJSON, err = stdlibjson.MarshalIndent(info, "", "    ")
```

Comment:
this is not needed colorjson turns-off if Iterminal is not detected you only need to do json.MarshalIndent @BigUstad

Reply:
Also you don't need to write this type of function but implement a function which implements JSON() which returns the json marshal of the `lifecycleConfiguration`

Reply:
Look at `copyMessage` implementation for String() and JSON() we should bring such implementation.

Reply:
> this is not needed colorjson turns-off if Iterminal is not detected you only need to do json.MarshalIndent @BigUstad
colorjson (json.MarshalIndent) was losing the padding/formatting - newlines, tabbed placements of elements et.al. when the output was redirected to a file. Hence, stdlibjson is used for that context.